[WIP] Rhar768/feature/23 rest interface

rowanh22 opened the pull request:
Refs https://github.com/SE761Team4/jabref/issues/23

rmoradc committed:
Add mindmap node data model

Code will need refactoring in terms of its package once parser and REST interface set up.

rmoradc committed:
Add parser class

rmoradc committed:
Add mind map edge data model

rmoradc committed:
Add model for full mind map

rmoradc committed:
Merge branch 'rmoradc/feature/mindmap-model' into rmoradc/feature/bibtext-mindmap-parser

rmoradc committed:
Rename class

rmoradc committed:
Add logic to parse BibtexString objects to MindMap objects

rowanh22 committed:
Added entry point for JabMap

koppor committed:
Initial (non-working) Root Resource

rowanh22 committed:
Refactored server creation (working)

rowanh22 committed:
Merge

rowanh22 committed:
Basic getting bib entry

rmoradc committed:
Merge branch 'rhar768/feature/23-REST-interface' into rmoradc/feature/38-bibtext-mindmap-parser

rmoradc committed:
Refactor to fit REST setup

This commit also fixes code style issues

rowanh22 committed:
Updated entries endpoint to return list of bib entries

rmoradc committed:
Update data object models

rmoradc committed:
Fix formatting

koppor commented on src/main/java/org/jabref/rest/resources/RootResource.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.ArrayList;
+import java.util.List;
+
+@Path("/root")
```
Is it possible to omit `/root`?

```suggestion
@Path("/")
```

koppor commented on src/main/java/org/jabref/rest/resources/RootResource.java:
```diff
@@ -0,0 +1,32 @@
+public class RootResource {
+
+    @GET
+    @Path("entries")
```
Please foresee the possibilitiy to work on multiple libraries (`libraries/current`)

```suggestion
    @Path("libraries/current/entries")
```

koppor commented on src/main/java/org/jabref/rest/resources/RootResource.java:
```diff
@@ -0,0 +1,32 @@
+    @Path("entries")
+    @Produces(MediaType.APPLICATION_JSON)
+    public String getEntries() {
+        List<BibEntry> entries = MindMapManager.instance().getBibEntries();
```
This code is a bit strange. I would go to the current opened library directly instead of routing the request throught he `MindMapManager`. (Since one does not need to be that specific here, the more general "object" is enough)

koppor commented on src/main/java/org/jabref/model/entry/BibEntrySerializer.java:
```diff
@@ -0,0 +1,42 @@
+    public JsonElement serialize(BibEntry entry, Type typeOfSrc, JsonSerializationContext context) {
+        JsonObject obj = new JsonObject();
+
+        obj.addProperty("type", entry.getType().getName());
```
Please introduce a constant (`public static final String JSON_TYPE`)

General thing: We need to add two meta information to the JSON. We need to the bibtex content and the meta data type and key. They should not be mixed up. We need to consider that a user can add custom fields.

To be prepared for moree meta data, I would suggest a nested object

```json
{
  "bibtex_metadata": {
    "type": "article",
    "key": "the-citation-key"
  }
}
```

In case a user adds `bibtex_metadata` as custom field in the entry editor, then the thing breaks.

Alternative: Do not add special handling for  `citationkey`, because JabRef itself uses that field name.

Another thing:

Shouldn't one use `typedEntry.getTypeForDisplay()` for type writing (Source: org.jabref.logic.bibtex.BibEntryWriter#writeRequiredFieldsFirstRemainingFieldsSecond)

```java
  TypedBibEntry typedEntry = new TypedBibEntry(entry, bibDatabaseMode);
  typedEntry.getTypeForDisplay();
```

koppor commented on src/main/java/org/jabref/model/entry/BibEntrySerializer.java:
```diff
@@ -0,0 +1,42 @@
+        JsonObject obj = new JsonObject();
+
+        obj.addProperty("type", entry.getType().getName());
+        obj.addProperty("cite_key", entry.getCiteKeyOptional().orElse(""));
```
Difficult decision here. Alternatives are `null` or to omit the field. It is OK for me to use the emptry string. Seems to be the JavaScript way.

koppor reviewed (commented):
Minor comments.

Please ensure that CheckStyle (and IntelliJ) is correctly setup at your side. Then, you can press <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>L</kbd> to fix the formatting (and <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> "Optimize Imports). One "annoying" thing of JabRef's checkstyle config is that there is a space needed after comment start (OK: `// comment`, not OK: `//comment`)

If possible, add Tests for the serialization. One can press <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>T</kbd> to quickly create test case skelletons.

See org.jabref.logic.bibtex.BibEntryWriterTest for an example test for BibEntry writing.

rmoradc committed:
Remove obselete MindMap class

rmoradc committed:
Add GET method for maps

rmoradc committed:
Add bibtex to mindmap adpater

rmoradc committed:
Add enums for node and edge fields

rmoradc committed:
Refactor data models with enums

rmoradc committed:
Add separate methods for bib to node and edge

MindMapNodes no longer keep list of children but have x and y pos. Done to prevent cyclic dependency complications when serialising.

rowanh22 committed:
Refactored bib entry serializer format & formatting

rowanh22 committed:
Converted node id to type Long rather than String

rowanh22 committed:
Merge branch 'rmoradc/feature/bibtext-mindmap-parser' into rhar768/feature/23-REST-interface

# Conflicts:
#	src/main/java/org/jabref/gui/jabmap/JabMapAction.java
#	src/main/java/org/jabref/logic/jabmap/MindMapManager.java
#	src/main/java/org/jabref/rest/resources/RootResource.java

rowanh22 committed:
Merged branch rmoradc/feature/bibtext-mindmap-parser and added basic get map endpoint [WIP]

rowanh22 committed:
Added reading mindmap from bibtex file and endpoint to get the current map

koppor commented:
Follow-up at https://github.com/SE761Team4/jabref/pull/76.
